### PR TITLE
Added a new fieldSplit preconditioner

### LINKED ISSUE #7261: Add a fieldSplit preconditioner
### FieldSplit preconditioner

The basic idea of fieldSplit preconditioner is to decouple the system into several subproblems such that each subproblem can be effectively solved using existing preconditioners such as ASM, AMG, ILU, etc. The decomposition could be based on physics or others.

There already exists an incomplete implementation in MOOSE, but there are some issues:
(1) The preconditioner is constructed in an independent block [Splits]. The fieldSplit should be one subclass of the moose preconditioner.
(2) PETSc options for subproblems can not be correctly handled.
(3) DMMoose gets reconstructed during each Newton iteration. It is not efficient. DMMoose should be rebuilt only when we have to, for example, underlying nonlinear systems changed.

### PR DESCRIPTION
**Pull request by fdkong**
(1) The old [Splits] block was permanently removed because we really do not want the fieldSplit as an independent block.

(2) Reorganized DMMoose a little bit, and fixed issues regarding applying prefix for subsolvers. PETSc options are stored in FEProblem now, and do not insert PETSc options separately.

Closes #7261

### CODE REVIEW

File: test/tests/preconditioners/fsp/sbp_test.i [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+  [./PBP]
+    type = PBP
+    solve_order = 'u v'
+    preconditioner  = 'LU LU'
+    off_diag_row    = 'v'
+    off_diag_column = 'u'
+
+    petsc_options = ''  # Test petsc options in PBP block
+  [../]
```
Comment (jwpeterson) [inline test/tests/preconditioners/fsp/sbp_test.i]: Please remove PBP section if you are not using it.

File: test/tests/preconditioners/fsp/sbp_test.i [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+  [./FSP]
```
Comment (jwpeterson) [inline test/tests/preconditioners/fsp/sbp_test.i]: Add a comment here describing the linear algebra of this preconditioner so that people know how it works...

**Comment by jwpeterson** [comment]
It appears that this patch contains mostly formatting changes, but @fdkong tells me that it fixes an actual bug as well... I've asked him to split up the bugfix and the formatting changes into separate patches to make reviewing easier.

**Comment by friedmud** [comment]
Cool - I like the new syntax - looks good!

File: framework/src/utils/PetscSupport.C [hunk 3]
```diff
@@ -136,16 +137,26 @@ setSolverOptions(SolverParams & solver_params)
   }
 }
 
-void petscSetupDM (NonlinearSystem & nl) {
+int petscSetupDM (NonlinearSystem & nl) {
```
Comment (jwpeterson) [inline framework/src/utils/PetscSupport.C]: Why did you change this function to return an `int`?  It never returns anything other than zero...
